graphql/ratelimit.go

```diff
@@ -0,0 +1,172 @@
+	// requests to databases
+	currentMaxRequestsLevel int
+	// predictedDuration a duration that could take current request. Calculation
+	// depends on real duration of previous requests
+	predictedDuration time.Duration
+	// if simultaneous requests of service reach level of
+	// currentMaxRequestsLevel service will wait `waitTime` before repeating
+	// attempt to serve request
+	waitTime time.Duration
+	mux      sync.Mutex
+}
+
+// activeRequest provides structure for request that processing by service
+// stores id, count of simultaneous requests on moment of starting request,
+// date of request start and predicted request duration
+type activeRequest struct {
+	id          uuid.UUID
+	bucketLevel int
+	startedAt   time.Time
+	predictedAt time.Time
+}
+
+// RatelimitHandlerDefault creates ratelimit object with empty values
+func RatelimitHandlerDefault() *RatelimitObject {
+	rObj := &RatelimitObject{}
+	return rObj.initiate()
+}
+
+// RatelimitHandler creates ratelimit struct object
+func RatelimitHandler(maxRequests int, minRequests int, waitTime time.Duration) *RatelimitObject {
+	rObj := RatelimitObject{
+		maxRequests:             maxRequests,
+		minRequests:             minRequests,
+		currentMaxRequestsLevel: maxRequests,
+		waitTime:                waitTime,
+		predictedDuration:       waitTime,
+	}
+	return rObj.initiate()
+}
+
+// initiate helps initialize ratelimitObject with default values
+func (rObj *RatelimitObject) initiate() *RatelimitObject {
+	rObj.mux.Lock()
+	if rObj.waitTime == 0 {
+		rObj.waitTime = waitTimeDefault
+	}
+	if rObj.predictedDuration == 0 {
+		rObj.predictedDuration = rObj.waitTime
+	}
+	if rObj.maxRequests == 0 {
+		rObj.maxRequests = maxSimultaneousRequestsDefault
+		rObj.currentMaxRequestsLevel = maxSimultaneousRequestsDefault
+	}
+	if rObj.minRequests == 0 {
+		rObj.minRequests = minSimultaneousRequestsDefault
+	}
+	rObj.mux.Unlock()
+	return rObj
+}
+
+// newRequest creates activeRequest object. Method grants that request will be
+// served
+func (rObj *RatelimitObject) newRequest() (uuid.UUID, error) {
+	connID := uuid.NewV4()
+	predictedAt := time.Now().Add(rObj.predictedDuration)
+	rObj.mux.Lock()
+	rObj.requestsBucket = append(rObj.requestsBucket, activeRequest{
+		id:          connID,
+		bucketLevel: len(rObj.requestsBucket) + 1,
+		startedAt:   time.Now(),
+		predictedAt: predictedAt,
+	})
+	rObj.mux.Unlock()
+	return connID, nil
+}
+
+// ServeRequest decides put request into work or not. If amount of simultaneous
+// requests is more than allowed (`currentMaxRequestsLevel` level), method will
+// wait for up to `httpHandler.waitTime` and repeat attempt to start request. If
+// request started: returns UUID of request, otherwise returns empty UUID and
+// error
+func (rObj *RatelimitObject) ServeRequest(isInitial bool) (uuid.UUID, error) {
+	if len(rObj.requestsBucket) < rObj.currentMaxRequestsLevel {
+		return rObj.newRequest()
+	}
+	if isInitial && rObj.predictedDuration <= rObj.waitTime {
```

Comment:
Read access to `predictedDuration` will race with a concurrent update in `EndRequest`.